src/realm/sync/network/default_socket.cpp

```diff
@@ -491,70 +492,69 @@ void DefaultSocketProvider::start()
 {
     std::unique_lock<std::mutex> lock(m_mutex);
     // Has the thread already been started or is running
-    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running)
+    if (m_state == State::Starting || m_state == State::Running)
         return; // early return
 
     // If the thread has been previously run, make sure it has been joined first
-    if (m_state == State::Stopping || m_state == State::Stopped) {
-        if (m_thread.joinable()) {
-            lock.unlock();
-            m_thread.join();
-            lock.lock();
-        }
+    if (m_state == State::Stopping) {
+        state_wait_for(lock, State::Stopped);
     }
 
     m_logger_ptr->trace("Default event loop: start()");
     REALM_ASSERT(m_state == State::Stopped);
-    do_state_update(State::Starting);
+    do_state_update(lock, State::Starting);
     m_thread = std::thread{&DefaultSocketProvider::event_loop, this};
-    lock.unlock();
     // Wait for the thread to start before continuing
-    state_wait_for(State::Started);
```

Comment:
I think this was actually causing some subtle races where we'd wait for the thread to reach the "Started" state, but it wouldn't actually be ready to serve work until the the "Running" state. Since the Started/Running states should always be a matter of microseconds apart, I just removed the "Started" state and made it so we don't enter the "Running" state until we're actually inside the event loop.